### PR TITLE
Standardise errors

### LINKED ISSUE #1282: Better error messages when tasks fails
We print the `error.message` part when tasks fails. This is not always very informative.

Case in point, if we have an exec module that runs the command `[mkdir, foo]` where the dir `foo` already exists, we get the following error:

```
Failed building my-module. Here is the output:
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Command failed with exit code 1 (EPERM): mkdir foo
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

1 build task(s) failed!
```

...which doesn't tell us a lot.

 If we however print `error.all` (assuming the error originated in an `execa` call, `all` is basically `stdout + stderr`), we get:

```
Failed building my-module. Here is the output:
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
mkdir: foo: File exists
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

1 build task(s) failed!
```

...which is way more helpful.

The underlying issue is that we use different libraries to run these kind of commands and they return error objects of different shapes. We should aim to standardise these and include the `stderr` part when applicable.

### PR DESCRIPTION
**Pull request by eysi09**
**What this PR does / why we need it**:

Previously, the error message was not visible when commands were executed via `execa` (see #1282 for an example). Furthermore `execa` and `spawn` would return different looking errors.

The `execa` error message looked like this:

Screenshot 2019-10-28 at 15 04 45

Now both `execa` and `spawn` errors look like this:

Screenshot 2019-10-28 at 15 06 01

In particular, the error itself is visible and the command and error are bold. (EDIT: Text is no longer bold, see comments below).

**Which issue(s) this PR fixes**:

Fixes #1282, #1283 

**Special notes for your reviewer**:

I updated the commit to address an issue that just came up (#1283). I explain the changes in the comments below so that we can compare the different screenshots.

I suggest looking at the final changes as opposed to commit by commit. The tests should also paint a good picture.

Also, I'm flagging this as a draft PR until we squash and merge #1262 as it depends on it.

### CODE REVIEW

**Comment by eysi09** [comment]
Actually, it might not be a good idea to include ansi characters in the message itself. Although the bold looks pretty good imo. Here's what it looks now:

Screenshot 2019-10-28 at 21 52 48

**Comment by eysi09** [comment]
Updated so that it only prints "Here are the last lines..." part if they are different from the error.

Also addressed the comments in #1283. Results in the following three changes:

1. "Unable to connect to cluster..." is set as the state of the active log line.
2. The "Here's the output:" part now contains the actual error (as opposed to the text from 1.)
3. The "Failed resolving one or more providers..." message now only lists the names of the providers but not the message itself since that's printed above. The error object still contains all the error messages.

So we go from this:

Screenshot 2019-10-28 at 22 37 09

...to this:

Screenshot 2019-10-28 at 22 36 16

File: garden-service/src/plugins/kubernetes/namespace.ts [hunk 1]
```diff
@@ -116,21 +117,11 @@ export async function prepareNamespaces({ ctx, log }: GetEnvironmentStatusParams
...
-      {
-        providerConfig: k8sCtx.provider.config,
-        message,
-      },
-    )
+    log.setError(dedent`
+      Unable to connect to Kubernetes cluster. Please make sure it is running, reachable and that you have the right context configured.
+    `)
+
+    throw new DeploymentError(err.message, { providerConfig: k8sCtx.provider.config })
```
Comment (edvald) [inline garden-service/src/plugins/kubernetes/namespace.ts]: This seems slightly odd to me. I'd rather suggest prepending the "Unable to connect" string to `err.message`.

Comment (edvald) [reply garden-service/src/plugins/kubernetes/namespace.ts]: I'd go with 3) 👍

Comment (eysi09) [reply garden-service/src/plugins/kubernetes/namespace.ts]: Would we not want to set some error message on the active log line though?

File: garden-service/test/unit/src/util/util.ts [hunk 2]
```diff
@@ -1,15 +1,159 @@
...
+          error: "ls: scottiepippen: No such file or directory",
+        }))
+      }
+    })
+  })
+
+  describe("spawn", () => {
+    before(function() {
+      // These tests depend on the underlying OS and are only executed on macOS
+      if (process.platform !== "darwin") {
```
[the remark this answers is not in the document]
Comment (eysi09) [reply garden-service/test/unit/src/util/util.ts]: One way to find out.

File: garden-service/src/plugins/kubernetes/namespace.ts (hunk 1 above)
Comment (edvald) [reply garden-service/src/plugins/kubernetes/namespace.ts]: We could simply set it to "Error" imo... Or something like that. "Failed" maybe.

File: garden-service/test/unit/src/util/util.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/test/unit/src/util/util.ts]: Didn't work directly, but you can modify the assert to match part of the string. That should be portable.

Another option would be a command that we control more explicitly, e.g. `sh -c "echo foo; exit 1"`. That way we know the exact string and exit code.

File: garden-service/src/plugins/kubernetes/namespace.ts (hunk 1 above)
Comment (eysi09) [reply garden-service/src/plugins/kubernetes/namespace.ts]: I set the error in the log line to "Unable to connect to cluster" as well.

Screenshot 2019-11-01 at 08 34 58

This looks fine IMO but happy to change if you have any better suggestions.

**Comment by eysi09** [comment]
@edvald  I've addressed the comments and the tests are passing.

File: garden-service/src/plugins/kubernetes/namespace.ts (hunk 1 above)
Comment (edvald) [reply garden-service/src/plugins/kubernetes/namespace.ts]: That looks okay but my concern with the inline error message is that it's provider-specific. Multiple providers might fail, and we should expect users to have more and more providers over time.

Comment (eysi09) [reply garden-service/src/plugins/kubernetes/namespace.ts]: Yeah I was kind of thinking that other providers would then just overwrite the error. But you're right, I'll use something more generic.

**Pull request merged**